sync15-adapter/src/sync_stateful.rs

```diff
@@ -0,0 +1,126 @@
+    client_init: Sync15StorageClientInit,
+    // the client we will reuse if possible.
+    client: Sync15StorageClient,
+}
+
+// This is used only to persist a GlobalState, which is serializable.
+
+// XXX - should we define this simply as a Cell<>, like last_client_info,
+// and leave the persisting to whoever provides the cell?
+
+// XXX - alternatively/additionally, instead of defining this in terms of
+// the GlobalState structure, we could consider defining it in terms of
+// Serializable/Deserializable?
+pub trait GlobalStateProvider {
+    fn load(&self) -> result::Result<Option<GlobalState>, failure::Error>;
+
+    fn save(&self, state: Option<&GlobalState>) -> result::Result<(), failure::Error>;
+}
+
+pub fn sync_stateful(
+    store: &Store,
+    gsp: &GlobalStateProvider,
+    last_client_info: &Cell<Option<ClientInfo>>,
+    storage_init: &Sync15StorageClientInit,
+    root_sync_key: &KeyBundle
+) -> result::Result<(), Error> {
+    let maybe_global = gsp.load()?;
+    // Note: We explicitly write a None back as the state, meaning if we
+    // unexpectedly fail below, the next sync will redownload meta/global,
+    // crypto/keys, etc. without needing to. Apparently this is both okay
+    // and by design.
+    gsp.save(None)?;
+    let mut global_state = match maybe_global {
+        Some(g) => g,
+        None => {
+            info!("First time through since unlock. Creating default global state.");
+            last_client_info.replace(None);
+            GlobalState::default()
+        }
+    };
+
+    // Ditto for the ClientInfo - if we fail below the GlobalStateProvider will
+    // not have the last client and client_init, so will be re-initialized on
+    // the next sync.
+    let client_info = match last_client_info.replace(None) {
+        Some(client_info) => {
+            if client_info.client_init != *storage_init {
+                ClientInfo {
+                    client_init: storage_init.clone(),
+                    client: Sync15StorageClient::new(storage_init.clone())?,
+                }
+            } else {
+                // we can reuse it.
+                client_info
+            }
+        },
+        None => {
+            ClientInfo {
+                client_init: storage_init.clone(),
+                client: Sync15StorageClient::new(storage_init.clone())?,
+            }
+        }
+    };
+
+    // Advance the state machine to the point where it can perform a full
+    // sync. This may involve uploading meta/global, crypto/keys etc.
+    {
+        // Scope borrow of `sync_info.client`
+        let mut state_machine =
+            SetupStateMachine::for_full_sync(&client_info.client, &root_sync_key);
+        info!("Advancing state machine to ready (full)");
+        global_state = state_machine.to_ready(global_state)?;
+    }
+
+    // Reset our local state if necessary.
+    if global_state.engines_that_need_local_reset().contains(&store.collection_name()) {
```

Comment:
Nit: `collection_name()` returns a new heap allocated string every time, we should avoid calling it more than necessary. Please store it in a variable and reuse it (and below).

Alternatively, we could probably `Store::collection_name()` return a `&'static str` which would avoid this, at the cost of some generality, but whatever.